modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java

```diff
@@ -49,35 +48,45 @@
 
     private final String rtnTypeStr;
     public final String name;
-    private final List<String> paramTypeStrs;
-    private final List<String> paramNameStrs;
     private final SBlock block;
     public final boolean synthetic;
 
+    final List<DType> paramTypes = new ArrayList<>();
+    private final List<String> paramNames = new ArrayList<>();
+
     private CompilerSettings settings;
 
     Class<?> returnType;
-    List<Class<?>> typeParameters;
     MethodType methodType;
 
     org.objectweb.asm.commons.Method method;
     List<Parameter> parameters = new ArrayList<>();
 
     private Variable loop = null;
 
-    public SFunction(Location location, String rtnType, String name,
-                     List<String> paramTypes, List<String> paramNames, SBlock block,
-                     boolean synthetic) {
+    public SFunction(Location location, String rtnType, String name, SBlock block, boolean synthetic) {
         super(location);
 
         this.rtnTypeStr = Objects.requireNonNull(rtnType);
         this.name = Objects.requireNonNull(name);
-        this.paramTypeStrs = Collections.unmodifiableList(paramTypes);
-        this.paramNameStrs = Collections.unmodifiableList(paramNames);
         this.block = Objects.requireNonNull(block);
         this.synthetic = synthetic;
     }
 
+    public void replace(ANode original, ANode target) {
+        int index = paramTypes.indexOf(original);
```

Comment:
couldn't there be multiple params of the same type, but this would always replace the first one?

Reply:
Currently, equals is not implemented so it would be a direct instance comparison. In the future we should protect this by having our own contains method to go along with replace/getChildren/etc.